src/model/guild/automod.rs

```diff
@@ -0,0 +1,859 @@
+//! Auto moderation types
+//!
+//! [Discord docs](https://discord.com/developers/docs/resources/auto-moderation)
+
+use std::fmt;
+
+use serde::de::{Deserializer, Error, IgnoredAny, MapAccess};
+use serde::ser::{SerializeStruct, Serializer};
+use serde::{Deserialize, Serialize};
+use serde_value::{DeserializerError, Value};
+
+use crate::model::id::{ChannelId, GuildId, MessageId, RoleId, RuleId, UserId};
+
+/// Configured auto moderation rule.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Rule {
+    /// ID of the rule.
+    pub id: RuleId,
+    /// ID of the guild this rule belonfs to.
+    pub guild_id: GuildId,
+    /// Name of the rule.
+    pub name: String,
+    /// ID of the user which created the rule.
+    pub creator_id: UserId,
+    /// Event context in which the rule should be checked.
+    pub event_type: EventType,
+    /// Characterizes the type of content which can trigger the rule.
+    pub trigger: Trigger,
+    /// Actions which will execute when the rule is triggered.
+    pub actions: Vec<Action>,
+    /// Whether the rule is enabled.
+    pub enabled: bool,
+    /// Roles that should not be affected by the rule.
+    ///
+    /// Maximum of 20.
+    pub exempt_roles: Vec<RoleId>,
+    /// Channels that should not be affected by the rule.
+    ///
+    /// Maximum of 50.
+    pub exempt_channels: Vec<ChannelId>,
+}
+
+impl<'de> Deserialize<'de> for Rule {
```

Comment:
Why is a manual Deserialization implementation needed?

Reply:
it's needed to combine the `trigger_type` and `trigger_metadata` fields into the `Trigger` enum.

Reply:
`#[serde(flatten)]` may be the solution

Reply:
Okay so `#[serde(flatten)]` on the trigger field and `#[serde(tag = "trigger_type", content = "trigger_metadata")]` ([adjacently tagged enum](https://serde.rs/enum-representations.html#adjacently-tagged)) on the Trigger enum _would_ have been the perfect solution _but_ serde still doesn't support integer tags [cuz dtolnay is radio silent on https://github.com/serde-rs/serde/pull/2056 :( ]

Reply:
I'm hesitated to use `#[serde(flatten)]` because its open issues and internal buffering.

Reply:
What open issues and internal buffering are you referring to?

Reply:
Just what I have seen in general and the inactivity in that area over time.
https://github.com/serde-rs/serde/search?q=flatten+enum&type=issues

Reply:
Seems to work fine here though https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b66f9cf04363bcd31a3c3c3f66c325aa